### ISSUE TITLE
incorrect detection of emacs per-file "mode line"

### ISSUE THREAD

**Issue by tesch1** [github/linguist#2230]
Many c++ library files have no extension.
The way to tell emacs to parse a file as c++ is to add in-file variable specifications, such as:
// -*- c++ -*-
or
// -*- C++ -*-
or
// -*- mode: c++ -*-
or
// -*- mode: C++ -*-
or
/* Local Variables: */
/* mode:c++         */
/* comment-column:0 */
/* End:             */

An example of one very common usage: https://github.com/google/libcxx/blob/master/include/iostream

linguist doesn't recognize these correctly, rather it seems to want a line like this:
// -*- mode: cpp -*-
which is wrong, because "cpp" is not a valid emacs major mode, and will not work in emacs.  Linguist is also unable to recognize the common short-hand:
// -*- c++ -*-
and even the non-sensical (but conforming to linguist's mistaken assumption that there's an emacs mode called 'cpp') doesn't work:
// -*- cpp -*-

For a description of the two methods that Emacs reads file modes (local variables, actually) from a file, please see:
http://ergoemacs.org/emacs_manual/emacs/Specifying-File-Variables.html

For a description of the general mechanism that Emacs (and Emacsen) use to choose a major mode: http://ergoemacs.org/emacs_manual/emacs/Choosing-Modes.html

Sorry I dont have time to fix this with a push request, but I hope a clear description of the issue helps.

**Comment by svnpenn** [comment]
You should link to the emacs spec that says `cpp` is incorrect and `c++` is correct, otherwise this is a feature request and not an error.

**Comment by tesch1** [comment]
There is no emacs mode called cpp.  Basically, the regex for grokking emacs file-specific variables is wrong, see my other comment on "Emacs vim modelines" for more info, and a fix.

**Comment by svnpenn** [comment]
Again, if you do not have a reference, it is just you saying it

**Comment by pchaigno** [comment]
> because "cpp" is not a valid emacs major mode, and will not work in emacs.

@svnpenn If you have Emacs that should be easy to verify, no?

Anyway, sounds like it wouldn't hurt to add support for `c++` modelines.
@tesch1 Do you want to open a pull request with that fix?

**Comment by larsbrinkhoff** [comment]
There is a bit of a disconnect between Emacs modelines and Linguist languages.  E.g. I don't think `-*- fundamental -*-` would be handled wery well.

**Comment by svnpenn** [comment]
another question, how does C++ highlighting differ from C highlighting? Surely they cant be that different?

**Comment by tesch1** [comment]
@larsbrinkhoff yeah, I mentioned that in the other thread.  In the pull-request I've added an alias for text mode called "fundamental" which seems to "do the right thing."  I guess most of the other major mode names should be close enough to be captured properly, maybe a few stragglers, but I guess if people need them they can fix them with fundamental as an example.

@pchaigno ok, I opened a pull request.  Unfortunately there's an unrelated test failing now because one of the DTrace test files has an emacs specification line in it that says to format the file as c++ (samples/DTrace/javascript-trace.d).  What's the right thing to do here, change the DTrace test file maybe?

**Comment by larsbrinkhoff** [comment]
I'm the one who added the DTrace sample file.  I didn't give any thought to the modeline, so I'd say keep the file, but change the mode to... wait for it... `dtrace-script`.  See here: https://github.com/dotemacs/dtrace-script-mode

**Comment by larsbrinkhoff** [comment]
DTrace sample fix in #2253

**Comment by tesch1** [comment]
Awesome! Glad I could contribute a tiny little bit, you guys do all the real work!
